raft/src/test/java/org/apache/kafka/raft/KafkaRaftClientPreVoteTest.java

```diff
@@ -0,0 +1,543 @@
+ * limitations under the License.
+ */
+package org.apache.kafka.raft;
+
+import org.apache.kafka.common.Uuid;
+import org.apache.kafka.common.protocol.Errors;
+import org.apache.kafka.common.record.MemoryRecords;
+
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.params.ParameterizedTest;
+import org.junit.jupiter.params.provider.ValueSource;
+
+import java.util.Collections;
+import java.util.OptionalInt;
+import java.util.Set;
+
+import static org.apache.kafka.raft.KafkaRaftClientTest.randomReplicaId;
+import static org.apache.kafka.raft.KafkaRaftClientTest.replicaKey;
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertTrue;
+
+public class KafkaRaftClientPreVoteTest {
+    @ParameterizedTest
+    @ValueSource(booleans = {true, false})
+    public void testHandlePreVoteRequestAsFollowerWithElectedLeader(boolean hasFetchedFromLeader) throws Exception {
+        int localId = randomReplicaId();
+        int epoch = 2;
+        ReplicaKey otherNodeKey = replicaKey(localId + 1, true);
+        int electedLeaderId = localId + 2;
+        Set<Integer> voters = Set.of(localId, otherNodeKey.id(), electedLeaderId);
+
+        RaftClientTestContext context = new RaftClientTestContext.Builder(localId, voters)
+            .withElectedLeader(epoch, electedLeaderId)
+            .withKip853Rpc(true)
+            .build();
+
+        if (hasFetchedFromLeader) {
+            context.pollUntilRequest();
+            RaftRequest.Outbound fetchRequest = context.assertSentFetchRequest();
+            context.assertFetchRequestData(fetchRequest, epoch, 0L, 0);
+
+            context.deliverResponse(
+                fetchRequest.correlationId(),
+                fetchRequest.destination(),
+                context.fetchResponse(epoch, electedLeaderId, MemoryRecords.EMPTY, 0L, Errors.NONE)
+            );
+        }
+
+        // follower should reject pre-vote requests with the same epoch if it has successfully fetched from the leader
+        context.deliverRequest(context.preVoteRequest(epoch, otherNodeKey, epoch, 1));
+        context.pollUntilResponse();
+
+        boolean voteGranted = !hasFetchedFromLeader;
+        context.assertSentVoteResponse(Errors.NONE, epoch, OptionalInt.of(electedLeaderId), voteGranted);
+        context.assertElectedLeader(epoch, electedLeaderId);
+
+        // follower will transition to unattached if pre-vote request has a higher epoch
+        context.deliverRequest(context.preVoteRequest(epoch + 1, otherNodeKey, epoch + 1, 1));
+        context.pollUntilResponse();
+
+        context.assertSentVoteResponse(Errors.NONE, epoch + 1, OptionalInt.of(-1), true);
+        assertEquals(context.currentEpoch(), epoch + 1);
+        assertTrue(context.client.quorum().isUnattachedNotVoted());
+    }
+
+    @Test
+    public void testHandlePreVoteRequestAsFollowerWithVotedCandidate() throws Exception {
```

Comment:
What does "AsFollowerWithVotedCandidtate"? Follower and UnattachedVoted are mutually exclusive states in the KRaft implementation.

Reply:
hm, it should not be valid to have a leaderId _and_ votedKey today, but I think this may change with the introduction of ProspectiveState

the situation I was thinking of was the following:
```
UnattachedVoted (votedKey=node3, epoch=10) -> Prospective (votedKey=node3, epoch=10) 
- election timeout, transition from unattached to prospective

Prospective (votedKey=node3, epoch=10) -> Follower (leader=node1, votedKey=node3, epoch=10)
- received PreVote response denoting leader=node1 in epoch 10, should it retain the votedKey or just clear it?
```
I think we need to retain the votedKey, because if Follower then transitions to Prospective again, it will be able to incorrectly grant another standard (persisted) vote unless it has a votedKey.

Reply:
this can be addressed in the next PR however, I'll remove this test

Reply:
> Prospective again, it will be able to incorrectly grant another standard (persisted) vote in epoch 10 unless it has a votedKey.

Let's talk about this in the next PR since this PR doesn't add that state. But yes, Prospective should not grant regular votes if the leader id is set. I guess this is similar to the unattached state which looks at both the leader id and voted key to determine if it should grant a vote.